### DISCUSSION TITLE
Upload only files type TXT, DAT or CSV

### DISCUSSION THREAD

**Discussion by patriciomh1965** [serenity-is/Serenity#6125] [Help]
Hello community
I need to upload TXT, DAT or CSV files. I couldn't find a way to do it.
Here I leave the instruction and image.

XYZForm
....
....
  [HalfWidth]
        [Required]
        [FileUploadEditor(FilenameFormat = "ArchivosManual/{0}_{4}", AllowNonImage = true, CopyToHistory = false)]
        public String EncArchivo { get; set; }
....
....
![archivos](https://user-images.githubusercontent.com/42218266/144512289-22d6da7d-4206-495b-9f65-943ee898e7aa.png)

I appreciate your help.
Atte.

**Comment by wezmag** [comment]
You can validate file extension in Create and Update method of the endpoint class.
For example:
``` csharp
public SaveResponse Create(IUnitOfWork uow, SaveRequest<MyRow> request)
{
    var uploadedFiles = JsonConvert.DeserializeObject<UploadedFile[]>(request.Entity.EncArchivo);
    foreach (var uploadedFile in uploadedFiles)
    {
        if (Path.GetExtension(uploadedFile.OriginalName).Equals(".csv", StringComparison.OrdinalIgnoreCase) == false)
        {
            throw new ValidationError("FileNameViolation", "EncArchivo", "Only csv files are allowed!");
        }
    }
    return new MyRepository(Context).Create(uow, request);
}
```

Please note that above code is not tested, you might need to modify it.
You can also do the validation in the repository class.

**Reply by patriciomh1965** [reply]
I appreciate your help, I will implement it

Thanks

**Reply by patriciomh1965** [reply]
Estimated

Can the file extension type be predefined in the FileUploadEditor object?

**Reply by wezmag** [reply]
I don't find any property for file extension in the FileUploadEditor. However, Serenity Platform is very open, you can always implement your own FileUploadEditor.
